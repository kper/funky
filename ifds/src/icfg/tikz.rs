use crate::counter::Counter;
use crate::icfg::graph::{Edge, Fact, Graph};
use crate::icfg::state::State;
use log::debug;
use std::cmp::Ordering;

const TAU: usize = 1;

struct FactWrapper<'a> {
    id: usize,
    fact: &'a Fact,
}

impl<'a> FactWrapper<'a> {
    pub fn get(&'a self) -> &'a Fact {
        self.fact
    }
}

/// Convert the given graph into string in .tex format.
pub fn render_to(graph: &Graph, state: &State) -> String {
    let mut str_vars = String::new();

    let mut index = 0;
    let mut functions: Vec<_> = state.functions.iter().map(|(_, f)| f).collect();

    functions.sort_by(|a, b| b.name.cmp(&a.name));

    let mut counter = Counter::default();

    let mut facts: Vec<_> = graph
        .edges
        .iter()
        .map(|x| x.get_from())
        .chain(graph.edges.iter().map(|x| x.to()))
        .collect();

    fn chain_ordering(o1: Ordering, o2: Ordering) -> Ordering {
        match o1 {
            Ordering::Equal => o1,
            _ => o2,
        }
    }

    facts.sort_by(|a, b| {
        chain_ordering(
            chain_ordering(
                b.function.cmp(&a.function),
                b.belongs_to_var.cmp(&a.belongs_to_var),
            ),
            b.next_pc.cmp(&a.next_pc),
        )
    });
    facts.dedup();

    let facts = facts
        .into_iter()
        .map(|x| FactWrapper {
            id: counter.get(),
            fact: x,
        })
        .collect::<Vec<_>>();

    for function in functions {
        let function_name = &function.name;
        debug!("Drawing function {}", function_name);

        let facts = facts.iter().filter(|x| &x.get().function == function_name);

        let notes = state.notes.iter().filter(|x| &x.function == function_name);

        let mut vars : Vec<_> = facts.clone().map(|x| &x.get().belongs_to_var).collect();
        vars.sort_by(|a, b| {
            b.cmp(a)
        });
        vars.dedup();

        for fact in facts {
            debug!("Drawing fact");

            str_vars.push_str(&format!(
                "\\node[circle,fill,inner sep=1pt,label=left:${}$] ({}) at ({}, {}) {{ }};\n",
                fact.get().belongs_to_var.replace("%", "\\%"),
                fact.id,
                index + vars.iter().position(|x| x == &&fact.get().belongs_to_var).unwrap(),
                fact.get().next_pc,
            ));
        }

        for note in notes {
            debug!("Drawing note");

            str_vars.push_str(&format!(
                "\\node[font=\\tiny] (note_{}) at ({}, {}) {{{}}};\n",
                note.id,
                index as f64 - 1.5,   //x
                note.pc as f64 + 0.5, //y
                note.note.replace("%", "").replace("\"", "").escape_debug(),
            ));
        }

        index += function.definitions + TAU + 1;
    }

    let get_fact_id = |fact| {
        let pos = facts.iter().position(|x| x.get() == fact).unwrap();
        facts.get(pos).unwrap().id
    };

    for edge in graph.edges.iter() {
        match edge {
            Edge::Normal { from, to, curved } => {
                if *curved {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->, bend right] ({}) edge ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                } else {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->] ({}) edge ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                }
            }
            Edge::Call { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, green] ({}) [bend left] edge node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
            Edge::CallToReturn { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->] ({}) edge node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
            Edge::Return { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend right] edge  node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
            Edge::Path { from, to } => {
                if from != to {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->, blue] ({}) [bend right] edge  node {{ }} ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                } else {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->, blue] ({}) [loop right] edge  node {{ }} ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                }
            }
            Edge::Summary { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend left] edge  node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
        }
    }

    template(str_vars)
}

fn template(inject: String) -> String {
    format!(
        "   \\documentclass{{standalone}}
            \\usepackage{{pgf, tikz}}
            \\usetikzlibrary{{arrows, automata}}
            \\begin{{document}}
                \\begin{{tikzpicture}}
                    [node distance=3cm]
                    \\tikzstyle{{every state}}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    {}

                \\end{{tikzpicture}}
            \\end{{document}}",
        inject
    )
}
