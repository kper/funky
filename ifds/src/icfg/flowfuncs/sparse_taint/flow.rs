use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
