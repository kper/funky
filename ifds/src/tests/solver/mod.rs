use crate::solver::*;

use crate::icfg::convert::ConvertSummary;
use crate::icfg::tikz::render_to;
use insta::assert_snapshot;

use crate::grammar::*;

use crate::icfg::flowfuncs::taint::flow::TaintNormalFlowFunction;
use crate::icfg::flowfuncs::taint::initial::TaintInitialFlowFunction;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {{
        let mut convert = ConvertSummary::new(TaintInitialFlowFunction, TaintNormalFlowFunction);

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let (graph, state) = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);

        graph
    }};
}

macro_rules! check_vars {
    ($name:expr, $solver:expr, $graph:expr, $req:expr) => {
        let mut vars = $solver
                    .sinks_var(&mut $graph, &$req)
                    .unwrap()
                    .into_iter()
                    .collect::<Vec<_>>();

        vars.sort();

        assert_snapshot!(
            format!("{}_vars", $name),
            format!(
                "{:#?}",
                vars 
            )
        );
    };
}

#[test]
fn test_intra_reachability() {
    let mut solver = IfdsSolver;

    let name = "intra_reach";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = %0
         };
    "
    );

    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_kill() {
    let mut solver = IfdsSolver;

    let name = "test_kill";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0) {
            %0 = 1
            KILL %0
         };
    "
    );

    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    // should not be reachable anymore
    assert!(!solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_loop() {
    let mut solver = IfdsSolver;

    let name = "test_loop";

    let req = Request {
        function: "main".to_string(),
        pc: 1,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 0 
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "main".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_functions() {
    let mut solver = IfdsSolver;

    let name = "test_functions";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0) (define %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_gcd() {
    let mut solver = IfdsSolver;

    let name = "test_gcd";

    let req = Request {
        function: "0".to_string(),
        pc: 17,
        variable: Some("%11".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
       define 0 (param %0 %1) (result 1) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16) {
        BLOCK 0
        BLOCK 1
        %4 = %0
        %5 = %1
        %6 = %5 op %4
        IF %6 THEN GOTO 3 ELSE GOTO 4
        BLOCK 3 
        %7 = %0
        %2 = %7
        KILL %7
        GOTO 2
        GOTO 4
        BLOCK 4
        BLOCK 5
        %8 = %1
        %9 = 0
        %10 = %0
        %11 = %1
        %12 = %11 op %10
        %13 = %12
        %2 = %12
        %14 = %13 op %12
        %3 = %14
        %15 = %1
        %16 = 0
        BLOCK 2
        RETURN %15;
        }; 
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_globals() {
    let mut solver = IfdsSolver;

    let name = "test_globals";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%-1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%2".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_returned_value() {
    let mut solver = IfdsSolver;

    let name = "test_returned_values";

    let req = Request {
        function: "0".to_string(),
        pc: 8,
        variable: Some("%11".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 1) (define %0 %1 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16 %17) {
            BLOCK 0
            BLOCK 1
            BLOCK 3
            %1 = %0
            %7 = %0
            %8 = -1
            %9 = %8 op %7
            %10 <- CALL 0(%9)
            %11 = %0
            %12 = -2
            %13 = %12 op %11
            %14 <- CALL 0(%13)
            %15 = %14 op %13
            KILL %15
            KILL %14
            RETURN %0;
        }; 
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_looped_param() {
    let mut solver = IfdsSolver;

    let name = "test_looped_param";

    let req = Request {
        function: "0".to_string(),
        pc: 1,
        variable: Some("%7".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 1) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9 %10 %11 %12 %13 %14 %15 %16 %17) {
            %1 = %0
            %7 = %0
            %9 = %8 op %7
            %10 <- CALL 0(%9)
            %11 = %0
            RETURN %0;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_memory() {
    let mut solver = IfdsSolver;

    let name = "test_memory";

    let req = Request {
        function: "0".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define 0 (param %0) (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
            %0 = -12345
            STORE FROM %0 OFFSET 0 + %0 ALIGN 2 32
            %4 = 8
            %5 = LOAD OFFSET 0 + %4 ALIGN 0
            %6 = 8
            %7 = LOAD OFFSET 0 + %6 ALIGN 0
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 0,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 1,
                variable: Some("%0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 1,
                variable: Some("mem@0".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 3,
                variable: Some("%5".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "0".to_string(),
                pc: 5,
                variable: Some("%7".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}

#[test]
fn test_ir_multiple_functions() {
    let mut solver = IfdsSolver;

    let name = "test_multiple_functions";

    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let mut graph = ir!(
        name,
        req,
        "
        define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0)  {
            RETURN %0;
        };
    "
    );

    let sinks = solver.all_sinks(&mut graph, &req).unwrap();

    assert_snapshot!(name, format!("{:#?}", sinks));

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 1,
                variable: Some("%1".to_string()),
            }
        )
        .unwrap());

    assert!(solver
        .is_taint(
            &mut graph,
            &req,
            &Request {
                function: "test".to_string(),
                pc: 2,
                variable: Some("%2".to_string()),
            }
        )
        .unwrap());

    check_vars!(name, solver, graph, req);
}
