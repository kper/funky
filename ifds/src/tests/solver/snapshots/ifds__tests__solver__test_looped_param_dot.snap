---
source: src/tests/solver/mod.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:$taut$] (0) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (1) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (2) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$\%0$] (3) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (4) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$\%0$] (5) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (6) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$\%0$] (7) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (8) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$\%0$] (9) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (10) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (11) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (12) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (13) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (14) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (15) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (16) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (17) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (18) at (2, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (19) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (20) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (21) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (22) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (23) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (24) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (25) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (26) at (1, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (27) at (2, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$\%0$] (28) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (29) at (1, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (30) at (2, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (31) at (4, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$\%0$] (32) at (5, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (33) at (1, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (34) at (2, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (35) at (4, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (36) at (2, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (37) at (1, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (38) at (2, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (39) at (4, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (40) at (1, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$\%7$] (41) at (2, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (42) at (3, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (43) at (4, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (44) at (3, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (45) at (4, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (46) at (3, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (47) at (4, 6) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Assign(1, 0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Assign(7, 0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {BinOp(9, 8, 7)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Call(0, [9], [10])};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Assign(11, 0)};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Return([0])};
\node[font=\tiny] (note_6) at (-1.5, 0.5) {Assign(1, 0)};
\node[font=\tiny] (note_7) at (-1.5, 1.5) {Assign(7, 0)};
\node[font=\tiny] (note_8) at (-1.5, 2.5) {BinOp(9, 8, 7)};
\node[font=\tiny] (note_9) at (-1.5, 3.5) {Call(0, [9], [10])};
\node[font=\tiny] (note_10) at (-1.5, 4.5) {Assign(11, 0)};
\node[font=\tiny] (note_11) at (-1.5, 5.5) {Return([0])};
\node[font=\tiny] (note_12) at (-1.5, 0.5) {Assign(1, 0)};
\node[font=\tiny] (note_13) at (-1.5, 1.5) {Assign(7, 0)};
\node[font=\tiny] (note_14) at (-1.5, 2.5) {BinOp(9, 8, 7)};
\node[font=\tiny] (note_15) at (-1.5, 3.5) {Call(0, [9], [10])};
\node[font=\tiny] (note_16) at (-1.5, 4.5) {Assign(11, 0)};
\node[font=\tiny] (note_17) at (-1.5, 5.5) {Return([0])};
\node[font=\tiny] (note_18) at (-1.5, 0.5) {Assign(1, 0)};
\node[font=\tiny] (note_19) at (-1.5, 1.5) {Assign(7, 0)};
\node[font=\tiny] (note_20) at (-1.5, 2.5) {BinOp(9, 8, 7)};
\node[font=\tiny] (note_21) at (-1.5, 3.5) {Call(0, [9], [10])};
\node[font=\tiny] (note_22) at (-1.5, 4.5) {Assign(11, 0)};
\node[font=\tiny] (note_23) at (-1.5, 5.5) {Return([0])};
\node[font=\tiny] (note_24) at (-1.5, 0.5) {Assign(1, 0)};
\node[font=\tiny] (note_25) at (-1.5, 1.5) {Assign(7, 0)};
\node[font=\tiny] (note_26) at (-1.5, 2.5) {BinOp(9, 8, 7)};
\node[font=\tiny] (note_27) at (-1.5, 3.5) {Call(0, [9], [10])};
\node[font=\tiny] (note_28) at (-1.5, 4.5) {Assign(11, 0)};
\node[font=\tiny] (note_29) at (-1.5, 5.5) {Return([0])};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (1);
		\path[->, blue] (0) [bend right] edge  node { } (13);
		\path[->, blue] (0) [bend right] edge  node { } (14);
		\path[->, blue] (0) [bend right] edge  node { } (15);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (0) [bend right] edge  node { } (17);
		\path[->, blue] (0) [bend right] edge  node { } (18);
		\path[->, blue] (1) [loop right] edge  node { } (1);
		\path[->, blue] (1) [bend right] edge  node { } (0);
		\path[->, blue] (1) [bend right] edge  node { } (13);
		\path[->, blue] (1) [bend right] edge  node { } (14);
		\path[->, blue] (1) [bend right] edge  node { } (15);
		\path[->, blue] (1) [bend right] edge  node { } (16);
		\path[->, blue] (1) [bend right] edge  node { } (17);
		\path[->, blue] (0) [bend right] edge  node { } (26);
		\path[->, blue] (0) [bend right] edge  node { } (27);
		\path[->, blue] (3) [loop right] edge  node { } (3);
		\path[->, blue] (1) [bend right] edge  node { } (29);
		\path[->, blue] (1) [bend right] edge  node { } (30);
		\path[->, blue] (3) [bend right] edge  node { } (31);
		\path[->, blue] (3) [bend right] edge  node { } (32);
		\path[->, blue] (1) [bend right] edge  node { } (33);
		\path[->, blue] (1) [bend right] edge  node { } (34);
		\path[->, blue] (3) [bend right] edge  node { } (35);
		\path[->, blue] (3) [bend right] edge  node { } (18);
		\path[->, blue] (1) [bend right] edge  node { } (37);
		\path[->, blue] (1) [bend right] edge  node { } (38);
		\path[->, blue] (3) [bend right] edge  node { } (39);
		\path[->, blue] (3) [bend right] edge  node { } (26);
		\path[->, blue] (3) [bend right] edge  node { } (27);
		\path[->, blue] (1) [bend right] edge  node { } (42);
		\path[->, blue] (1) [bend right] edge  node { } (43);
		\path[->, blue] (1) [bend right] edge  node { } (44);
		\path[->, blue] (1) [bend right] edge  node { } (45);
		\path[->, blue] (1) [bend right] edge  node { } (46);
		\path[->, blue] (1) [bend right] edge  node { } (47);


                \end{tikzpicture}
            \end{document}
